File: Matlab/Scripts/ex8_2_2.m
% exercise 8.2.1

% Load data
cdir = fileparts(mfilename('fullpath')); 
load(fullfile(cdir,'../Data/synth5'))

%% Fit model using boosting (AdaBoost)

% Number of rounds of boosting
L = 100;

% Allocate variable for model parameters
w_est = nan(M+1, L);

% Allocate variable for model importance weights
alpha = nan(1, L);

% Weights for selecting samples in each round of boosting
weights = ones(N,1)/N;
 
%% Boosting (AdaBoost)
% For each round of boosting
l = 1;
while l<=L
    % Choose data objects by random sampling with replacement 
    i = discreternd(weights, N);
%     for j=1:length(i)
%         i(j) = randsample(N,1,true,weights); 
%     end
     
    % Extract training set
    X_train = X(i, :);
    y_train = y(i);

    % Fit logistic regression model to training data and save result
    %if nnz(y_train == 0) == 0 || nnz(y_train == 1) == 0, 
    %    assert(false);
    %end
    %fprintf('%g, %g\n', nnz(y_train==0),nnz(y_train==1));
    w_est(:,l) = glmfit(X_train, y_train, 'binomial');
    %glmfit(X, y, 'binomial')
    % Make predictions on the whole data set
    p = glmval(w_est(:,l), X, 'logit');    
    
    y_est = p>0.5;
      
    % Compute error rate
    ErrorRate = sum(weights.*(y~=y_est))/N;
    
    %disp(sum(y~=y_est)/N)
    % Restart if error rate is above 50%
    % From Tan et al. p. 289: "In addition, if any intermediate rounds
    % produce an error rate higher than 50%, the weights are reverted back
    % to their original uniform values (...) and the resampling procedure
    % is repeated. 
    if ErrorRate > .5
        % Reset weights and do a new round
        weights = ones(N,1)/N;
        %disp('breaking, bad error rate');
        %disp(ErrorRate)
    else 
        % Compute model importance weight
        % Tan et al. p. 288, equation below equation 5.68
        alpha(l) = .5*log((1-ErrorRate)/ErrorRate);
 
        % Update weights
        % Tan et al. p. 288, equation 5.69
        weights(y==y_est) = weights(y==y_est)*exp(-alpha(l));
        weights(y~=y_est) = weights(y~=y_est)*exp(alpha(l));
        weights = weights/sum(weights);
        
        % Next round
        l = l+1;
    end    
end

% Normalize the importance weights
alpha = alpha/sum(alpha);

% Evaluate the logistic regression on the training data
p = glmval(w_est, X, 'logit');

% From Tan el al. p. 288: "Instead of using a majority voting scheme, the
% prediction made by each classifier (...) is weighted according to
% (alpha).
y_est = sum(bsxfun(@times, p>.5, alpha), 2)>.5;

% Compute error rate
ErrorRate = sum(y~=y_est)/N;
fprintf('Error rate: %.0f%%\n', ErrorRate*100);

%% Plot decision boundary
mfig('Decision boundary'); clf;
dbplot(X, y, @(X) sum(bsxfun(@times, glmval(w_est, X, 'logit')>.5, alpha), 2));
xlabel(attributeNames(1)); ylabel(attributeNames(2));
legend(classNames);

File: Python/Scripts/ex8_2_2.py
# exercise 8.2.2


from pylab import *
from scipy.io import loadmat
from toolbox_02450 import dbplot, dbprobplot, bootstrap
from bin_classifier_ensemble import BinClassifierEnsemble
from sklearn.linear_model import LogisticRegression

# Load Matlab data file and extract variables of interest
mat_data = loadmat('../Data/synth5.mat')
X = np.matrix(mat_data['X'])
y = np.matrix(mat_data['y'])
attributeNames = [name[0] for name in mat_data['attributeNames'].squeeze()]
classNames = [name[0][0] for name in mat_data['classNames']]
N, M = X.shape
C = len(classNames)


# Fit model using bootstrap aggregation (boosting, AdaBoost):

# Number of rounds of bagging
L = 100

# Weights for selecting samples in each bootstrap
weights = np.ones((N,1),dtype=float)/N

# Storage of trained log.reg. classifiers fitted in each bootstrap
logits = [0]*L
alpha = np.ones( (L,1) )
votes = np.zeros((N,1))
epsi = 0
y_all = np.zeros((N,L))
y = y > 0.5
# For each round of bagging
for l in range(L):
    while True : 
        # Extract training set by random sampling with replacement from X and y
        while True : 
            # not a thing of beauty, however log.reg. fails if presented with less than two classes. 
            X_train, y_train = bootstrap(X, y, N, weights) 
            if not (all(y_train==0) or all(y_train == 1)) : break      
        
        # Fit logistic regression model to training data and save result
        # turn off regularization with C. 
        logit_classifier = LogisticRegression(C=100000000)
    
        logit_classifier.fit(X_train, y_train.A.ravel()  )
        logits[l] = logit_classifier
        y_est = np.mat(logit_classifier.predict(X)).T > 0.5
        
        y_all[:,l] = 1.0 * y_est.ravel()
        v  = matrix(y_est != y,dtype=float)
        ErrorRate = multiply(weights,v).sum()/N
        epsi = ErrorRate
        if epsi > 0.5 : 
            weights = np.ones( (N,1), dtype=float)/N
            print "resetting weights..."
        else : 
            alphai = 0.5 * log( (1-epsi)/epsi)
            weights[y_est == y] = weights[y_est == y] * exp( -alphai )
            weights[y_est != y] = weights[y_est != y] * exp(  alphai )
            
            weights = weights / sum(weights)
            break;
            
    votes = votes + y_est
    alpha[l] = alphai
    print('Error rate: {0}%'.format(ErrorRate*100))    
    
# Estimated value of class labels (using 0.5 as threshold) by majority voting
alpha = mat(alpha)/sum(alpha)
y_est_ensemble = mat(y_all) * alpha > 0.5

#y_est_ensemble = votes > (L/2)
#y_est_ensemble = mat(y_all) * mat(alpha) - (1-mat(y_all)) * mat(alpha) > 0
ErrorRateEnsemble = sum(y_est_ensemble != y)/N

# Compute error rate
#ErrorRate = (y!=y_est_ensemble).sum(dtype=float)/N
print('Error rate for ensemble classifier: {:.1f}%'.format(ErrorRateEnsemble*100))

ce = BinClassifierEnsemble(logits)
figure(1); dbprobplot(ce, X, y, 'auto', resolution=200)
figure(2); dbplot(ce, X, y, 'auto', resolution=200)
figure(3); plot(alpha.A)

show()